src/GraphWebhooks/Controllers/ListenController.cs

```diff
@@ -0,0 +1,215 @@
+        /// <returns>IActionResult</returns>
+        [HttpPost]
+        [AllowAnonymous]
+        public async Task<IActionResult> Index([FromQuery] string validationToken = null)
+        {
+            // If there is a validation token in the query string,
+            // send it back in a 200 OK text/plain response
+            if (!string.IsNullOrEmpty(validationToken))
+            {
+                return Ok(validationToken);
+            }
+
+            // Read the body
+            using var reader = new StreamReader(Request.Body);
+            var jsonPayload = await reader.ReadToEndAsync();
+
+            // Use the Graph client's serializer to deserialize the body
+            var notifications = _graphClient.HttpProvider.Serializer
+                .DeserializeObject<ChangeNotificationCollection>(jsonPayload);
+
+            if (notifications == null) return Ok();
+
+            // Validate any tokens in the payload
+            bool areTokensValid = await AreTokensValid(notifications, _tenantIds, _appIds);
+            if (!areTokensValid) return Unauthorized();
+
+            // Process non-encrypted notifications first
+            // These will be notifications for user mailbox
+            var messageNotifications = new Dictionary<string, ChangeNotification>();
+            foreach (var notification in notifications.Value.Where(n => n.EncryptedContent == null))
+            {
+                // Find the subscription in our store
+                var subscription = _subscriptionStore
+                    .GetSubscriptionRecord(notification.SubscriptionId.ToString());
+
+                // If this isn't a subscription we know about, or if client state doesnt' match,
+                // ignore it
+                if (subscription != null && subscription.ClientState == notification.ClientState)
+                {
+                    _logger.LogInformation($"Received notification for: {notification.Resource}");
+                    // Add notification to list to process. If there is more than
+                    // one notification for a given resource, we'll only process it once
+                    messageNotifications[notification.Resource] = notification;
+                }
+            }
+
+            // Since resource data is not included in these notifications,
+            // use Microsoft Graph to get the messages
+            await GetMessagesAsync(messageNotifications.Values);
+
+            // Process encrypted notifications
+            var clientNotifications = new List<ClientNotification>();
+            foreach (var notification in notifications.Value.Where(n => n.EncryptedContent != null))
+            {
+                // Decrypt the encrypted payload using private key
+                var chatMessage = await notification.EncryptedContent.DecryptAsync<ChatMessage>(async (id, thumbprint) => {
```

Comment:
await async not needed here

Reply:
Assuming you mean in the lambda `certificateProvider` parameter.

Reply:
yes